Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -5384,6 +4290,241 @@ - (void)presentViewController:(UIViewController *)viewController
     [super presentViewController:viewController animated:animated completion:completion];
 }
 
+#pragma mark - ConversationViewModelDelegate
+
+- (void)conversationViewModelWillUpdate
+{
+    OWSAssertIsOnMainThread();
+    OWSAssertDebug(self.conversationViewModel);
+
+    OWSLogVerbose(@"");
+
+    // HACK to work around radar #28167779
+    // "UICollectionView performBatchUpdates can trigger a crash if the collection view is flagged for layout"
+    // more: https://github.com/PSPDFKit-labs/radar.apple.com/tree/master/28167779%20-%20CollectionViewBatchingIssue
+    // This was our #2 crash, and much exacerbated by the refactoring somewhere between 2.6.2.0-2.6.3.8
+    //
+    // NOTE: It's critical we do this before beginLongLivedReadTransaction.
+    //       We want to relayout our contents using the old message mappings and
+    //       view items before they are updated.
+    [self.collectionView layoutIfNeeded];
+    // ENDHACK to work around radar #28167779
+}
+
+- (void)conversationViewModelDidUpdate:(ConversationUpdate *)conversationUpdate
+{
+    OWSAssertIsOnMainThread();
+    OWSAssertDebug(conversationUpdate);
+    OWSAssertDebug(self.conversationViewModel);
+
+    if (!self.shouldObserveVMUpdates) {
+        return;
+    }
+
+    OWSLogVerbose(@"");
+
+    [self updateBackButtonUnreadCount];
+    [self updateNavigationBarSubtitleLabel];
+
+    if (self.isGroupConversation) {
+        [self.uiDatabaseConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
+            [self.thread reloadWithTransaction:transaction];
+        }];
+        [self updateNavigationTitle];
+    }
+
+    [self updateDisappearingMessagesConfiguration];
+
+    if (conversationUpdate.conversationUpdateType == ConversationUpdateType_Minor) {
+        return;
+    } else if (conversationUpdate.conversationUpdateType == ConversationUpdateType_Reload) {
+        [self resetContentAndLayout];
+        [self updateLastVisibleTimestamp];
+        [self scrollToBottomAnimated:NO];
+        return;
+    }
+
+    OWSAssertDebug(conversationUpdate.conversationUpdateType == ConversationUpdateType_Diff);
+    OWSAssertDebug(conversationUpdate.updateItems);
+
+    BOOL wasAtBottom = [self isScrolledToBottom];
+    // We want sending messages to feel snappy.  So, if the only
+    // update is a new outgoing message AND we're already scrolled to
+    // the bottom of the conversation, skip the scroll animation.
+    __block BOOL shouldAnimateScrollToBottom = !wasAtBottom;
+    // We want to scroll to the bottom if the user:
+    //
+    // a) already was at the bottom of the conversation.
+    // b) is inserting new interactions.
+    __block BOOL scrollToBottom = wasAtBottom;
+
+    void (^batchUpdates)(void) = ^{
+        OWSAssertIsOnMainThread();
+
+        const NSUInteger section = 0;
+        BOOL hasInserted = NO, hasUpdated = NO;
+        for (ConversationUpdateItem *updateItem in conversationUpdate.updateItems) {
```

Comment:
Much of this logic around handling incremental updates should look familiar.  The difference is that we now use `ConversationUpdateItem` instead of `YapDatabaseViewRowChange`.